### PR TITLE
Add extern module attribute

### PR DESCRIPTION
**Pull request by jedekar**

### CODE REVIEW

File: source/fe/ast/parser.d [hunk 1]
```diff
@@ -361,6 +369,16 @@ struct Parser
 				nextToken; // skip integer
 				attribute = make!BuiltinAttribNode(start, BuiltinAttribSubType.extern_syscall, syscallNumber);
 				break;
+			case CommonIds.id_module.index:
+				expectAndConsume(TokenType.COMMA, "@extern(module");
+				expect(TT.STRING_LITERAL, "@extern(module,");
+				string value = cast(string)context.getTokenString(tok.index);
+				//import std.algorithm.iteration : filter;
+				//uint syscallNumber = value.filter!(c => c != '_').to!uint;
```
Comment (MrSmith33) [inline source/fe/ast/parser.d:377]: This needs to be removed

File: source/fe/ast/parser.d [hunk 2]
```diff
@@ -361,6 +369,16 @@ struct Parser
...
+				nextToken; // skip lib name
+				//TODO: pass value as data
```
Comment (MrSmith33) [inline source/fe/ast/parser.d:379]: You can store intern module name in `context.idMap`, and store `Identifier` inside `data`.

File: source/fe/ast/decl/attribute.d [hunk 3]
```diff
@@ -65,6 +66,9 @@ void print_builtin_attribute(BuiltinAttribNode* node, ref AstPrintState state)
 		case BuiltinAttribSubType.extern_syscall:
 			state.print("ATTRIB @extern(syscall, ", node.data, ")");
 			break;
+		case BuiltinAttribSubType.extern_module:
+			state.print("ATTRIB @extern(module, ", node.data, ")");
```
Comment (MrSmith33) [inline source/fe/ast/decl/attribute.d:70]: Here you can print the `Identifier` via `context.idString`

**Review by MrSmith33 — CHANGES_REQUESTED** [review]
Can you also add a simple test case that checks that parsing passes with that attribute attached to the function?
You can copy-paste [this test](https://github.com/MrSmith33/vox/blob/master/source/tests/passing.d#L48-L62) to the end of the file with new test id.

**Comment by jedekar** [comment]
A few points I want to comment on:
- This copy-and-paste test tests only parsing, but I didn't see any attribute check for `syscall` case.
- Doesn't the addition of module name to identifiers corrupt `idMap`? For example, I want a function/variable `kernel32`, won't they conflict with each other?
- I changed the type of `data` to `Identifier`, and now `syscall` case uses `index` field as a syscall number, which is bad. I just think I misunderstood you somewhere. I thought about the union of `Identifier` and `uint` for `data`, but that's bad too.

File: source/fe/ast/parser.d [hunk 4]
```diff
@@ -359,7 +367,15 @@ struct Parser
...
-				attribute = make!BuiltinAttribNode(start, BuiltinAttribSubType.extern_syscall, syscallNumber);
+				attribute = make!BuiltinAttribNode(start, BuiltinAttribSubType.extern_syscall, Identifier(syscallNumber));
+				break;
+			case CommonIds.id_module.index:
+				expectAndConsume(TokenType.COMMA, "@extern(module");
+				expect(TT.STRING_LITERAL, "@extern(module,");
+				string value = cast(string)context.getTokenString(tok.index);
+				nextToken; // skip lib name
+				Identifier moduleId = context.idMap.getOrRegNoDup(context, value);
+				attribute = make!BuiltinAttribNode(start, BuiltinAttribSubType.extern_module, moduleId);
```
Comment (MrSmith33) [inline source/fe/ast/parser.d:378]: What I meant is that `Identifier.index` can be stored in `uint data` of `BuiltinAttribNode`.

File: source/tests/passing.d [hunk 5]
```diff
@@ -4737,3 +4737,21 @@ immutable test221 = q{--- test221
 	}
 	void debugCallback() {}
 };
+
+
+@TestInfo()
+immutable test222 = q{--- test222
+	@extern(module, "kernel32")
+	i32 test(i32 number) {
+		i32 result;
```
Comment (MrSmith33) [inline source/tests/passing.d:4746]: This should not need a body. But I should note to myself that directly applying `@extern` to functions with body should be an error.

Comment (jedekar) [reply source/tests/passing.d:4746]: Oh, that's a good test case scenario

**Comment by MrSmith33** [comment]
> A few points I want to comment on:
>     * This copy-and-paste test tests only parsing, but I didn't see any attribute check for `syscall` case.
Yea, the way it works right now would require generating an executable for linux, which I didn't do. Ideally needs a test too.
>     * Doesn't the addition of module name to identifiers corrupt `idMap`? For example, I want a function/variable `kernel32`, won't they conflict with each other?
No, because `idMap` is just used to intern strings for O(1) comparison on identifiers. Storing strings there is fine.
>     * I changed the type of `data` to `Identifier`, and now `syscall` case uses `index` field as a syscall number, which is bad. I just think I misunderstood you somewhere. I thought about the union of `Identifier` and `uint` for `data`, but that's bad too.
>       Isn't `data` supposed to be generic and low-level, like `[]u8` or something. It means we can place there anything, we just need a way for interpreting it.
Commented on that.

Forgot to mention, that you need to strip `"` on string literal, before putting it into `idMap`. Also, check the test output of `driver.context.printAstSema = true;` in tester. Did you figure out how to run single test?

File: source/tests/passing.d (hunk 5 above)
Comment (MrSmith33) [reply source/tests/passing.d:4746]: There is a file for tests that should fail

**Comment by jedekar** [comment]
Yeah, sure, I read the build/testing manual.  I didn't get to see the output for that exact test, because I corrected it. It brought another issue, though.

File: source/tests/passing.d (hunk 5 above)
Comment (jedekar) [reply source/tests/passing.d]: Got it

**Pull request merged**